Improved Selectable Properties

mcaskill opened the pull request:
Changes:
- Revised `SelectableProperty` mixin:
  - Added `addChoices()`, `hasChoices()`, `choiceLabel()`
  - Added `parseChoices()` and `parseChoice()` for use in subclasses
  - Cleaned up method documentation;
- `LangProperty`: Revised use of available locales, added support for customizing choices
- `ObjectProperty`:
  - Added `pagination` to further limit the number of selectable choices
  - Added `choiceLabel()`, `hasChoices()`, `parseChoices()`, `parseChoice()`, and exception for `addChoices()`
  - Moved preparation of collection loader (filters and orders) from `choices()` and `hasChoices()` to new `collectionModelLoader()`
  - Changed `modelLoader()` visibility from _private_ to _protected_
  - Modified `modelLoader()` to accept an optional object type 
    (e.g., useful for extended classes that might provide `<optgroup>` support)
- Updated unit tests

Highlights / (Possibly) Breaking Changes:
- Formatting of choice structure moved from `choice()` to `parseChoice()`.  
  _Update any custom selectable properties, accordingly._

mcaskill committed:
Improved Selectable Properties

Todo:
- ObjectProperty: Complete Unit Tests
- StringProperty: Complete Unit Tests

Changes SelectableProperty Mixin:
- Added ‘addChoices()’;
- Added ‘hasChoices()’;
- Added ‘choiceLabel()’;
- Added ‘parseChoices()’ and ‘parseChoice()’ for use in subclasses;
- Moved choice formatting from ‘addChoice()’ to ‘parseChoice()’;
- Cleaned up method documentation;

Changes:
- Moved choice label usage from ‘StringProperty::displayVal()’ to ‘AbstractProperty::displayVal()’ since its used by most that use ‘SelectablePropertyTrait’;
- ObjectProperty: Added ‘choiceLabel()’, ‘hasChoices()’, and exception for ’addChoices()’;
- LangProperty: Revised use of available locales, added support for customizing choices;
- Improved unit tests for the affected properties and mixins;

mducharme commented on src/Charcoal/Property/AbstractProperty.php:
```diff
@@ -431,6 +432,19 @@ public function displayVal($val, array $options = [])
             }
         }
 
+        if ($this instanceof SelectablePropertyInterface) {
```
Any ways this runtime check could be avoided?

mcaskill committed:
ObjectProperty: Modified model loader

Changes:
- Changed `modelLoader()` visibility from private to protected
- Added support for retrieving a specific model loader from `modelLoader()`

mcaskill committed:
ObjectProperty: Added Pagination

mcaskill committed:
ObjectProperty: Modified CollectionLoader Usage

Moved preparation of collection loader (filters and orders) from ‘choices()’ and ‘hasChoices()’ to ‘collectionLoader()’.

mcaskill committed:
StringProperty: Completed Unit Tests

Changes:
- Cleaned up `StringProperty`
- Completed unit tests for `StringProperty`
- Revised various method and property documentation
- Fixed 'SelectablePropertyTrait::parseChoice()' when receiving a `"key": "label"` structure
- Fixed elegant separators for displayed values
- Improved `ContainerIntegrationTrait`
- Added `ContainerProvider::registerMultilingualTranslator()`, used by `StringPropertyTest` and `LangPropertyTest`

mcaskill committed:
ObjectProperty: Completed Unit Tests

Changes:
- Cleaned up `ObjectProperty`
- Completed unit tests for `ObjectProperty`
- Implemented 'GenericModel' for testing
- Implemented "tedivm/stash" and "mustache/mustache" for unit tests
- Added "locomotivemtl/charcoal-view" services to `ContainerProvider`
- Fixed model collection loader service for unit tests
- Improved value display and pattern rendering for `ObjectProperty`
- Decoupled choice formatting from `ObjectProperty::choice()` to `ObjectProperty::parseChoice()`
- Minor cleanup of `LangPropertyTest`

mcaskill committed:
Fixed Scrutinizer Issues

mcaskill commented:
@mducharme Ready for review
